### ISSUE TITLE
progress_deadline distinguish between updates (auto_revert) and deployments

### ISSUE THREAD

**Issue by benbuzbee** [hashicorp/nomad#12940]
### Proposal
Add a new progress_deadline setting that only applies to updates and auto_revert decision making.

Today if you set progress_deadline to 5m and deploy a change to a job that, lets say, references a non-existing docker image, and set auto_revert to true; it will try for 5 minutes, fail (image not found), then revert to the old deployment.

This sounds good.

But what if, lets say the docker-registry is down for 15 minutes and you try to deploy a change?

I believe it will try for five minutes, fail, revert, continue to fail for five minutes (the problem wasnt the job, it was the docker-registry), and then give up...forever

Nomad giving up forever is a very bad and unexpected state. To get around it, maybe we want to set progress_deadline to 99999h (wish we could set unlimited = true here), but then you break the first case - auto_revert can never work.

So what I think I want is to tell nomad that if you are deciding whether to revert or not, consider the deadline 5m
But if you are deciding whether to fail the deployment, and wont revert, but will leave the job dead, consider a different deadline.

Why is the deadline for reverting the same as leaving the job dead forever?

**Comment by DerekStrickland** [comment]
Hi @benbuzbee 

I'm curious. Have you tried using [reschedule](https://www.nomadproject.io/docs/job-specification/reschedule) or [retry](https://www.nomadproject.io/docs/job-specification/restart) to achieve your desired behavior?

**Comment by benbuzbee** [comment]
> Hi @benbuzbee
> 
> I'm curious. Have you tried using [reschedule](https://www.nomadproject.io/docs/job-specification/reschedule) or [retry](https://www.nomadproject.io/docs/job-specification/restart) to achieve your desired behavior?

In my example, if the registry were down for 15 minutes
reschedule {
  unlimited = true
  }
  
  Does not prevent it from reaching the 5 minute progress_deadline, after which nomad will mark the deployment as failed and stop rescheduling it (even though we mark unlimited=true. To prove this, create a job that references a non existent docker image, set it to reschedule forever, with a progress deadline, and note that it eventually reaches the dead state and gives up.

Unlimited restarts might work, I did not test it necessarily, but I dont think thats a good solution as there are many reasons why keeping an allocation pinned to a node forever is undesirable

**Comment by DerekStrickland** [comment]
Got it!  Thanks for the extra context.

**Comment by axsuul** [comment]
Yep I'm facing a similar issue. We need a way to keep attempting to update indefinitely since it's much worse to just stop trying.
